File: LEETCODE/LineSweep/732_myCalendar3/732_myCalendar3.py

from collections import defaultdict
class MyCalendarThree:
    def __init__(self):
        self.delta= defaultdict(int)
        
    def book(self, startTime: int, endTime: int) -> int:
        self.delta[startTime]+=1
        self.delta[endTime]-=1
        ongoing,max_k=0,0
        for t in sorted(self.delta):
            ongoing+= self.delta[t]
            max_k= max(max_k,ongoing)
        return max_k

File: LEETCODE/LineSweep/732_myCalendar3/732_myCalendar3.java
class MyCalendarThree {
    TreeMap<Integer,Integer> map;
    public MyCalendarThree() {
        map= new TreeMap<>();
        
    }
    
    public int book(int startTime, int endTime) {
        map.put(startTime,map.getOrDefault(startTime,0)+1);
        map.put(endTime,map.getOrDefault(endTime,0)-1);
        int maxK=0;
        int ongoing=0;
        for(int delta: map.values()){
            ongoing+=delta;
            maxK= Math.max(maxK,ongoing);
        }
        return maxK;
    }
}